ledger-core/virtual/handlers/vstatereport.go

```diff
@@ -60,14 +60,86 @@ func (s *SMVStateReport) Init(ctx smachine.InitializationContext) smachine.State
 	}
 
 	ctx.SetDefaultMigration(s.migrationDefault)
-	return ctx.Jump(s.stepProcess)
+	return ctx.Jump(s.stepValidate)
 }
 
 func (s *SMVStateReport) migrationDefault(ctx smachine.MigrationContext) smachine.StateUpdate {
 	ctx.Log().Trace("stop processing VStateReport since pulse was changed")
 	return ctx.Stop()
 }
 
+func (s *SMVStateReport) stepValidate(ctx smachine.ExecutionContext) smachine.StateUpdate {
+	switch report := s.Payload; {
+	case report.GetObject().IsEmpty():
+		return ctx.Error(throw.New("Object should not be empty"))
+	case report.GetAsOf().IsUnknown():
+		return ctx.Error(throw.New("AsOf can not be Unknown"))
+	case !report.GetDelegationSpec().IsZero():
+		return ctx.Error(throw.New("DelegationSpec should be empty"))
+	default:
+		switch report.GetStatus() {
+		case payload.Unknown:
+			return ctx.Error(throw.New("Unexepected Unknown status received"))
+		case payload.Missing:
+			if err := validateZeroPending(report); err != nil {
+				return ctx.Error(err)
+			}
+
+			if err := validateEmptyState(report); err != nil {
+				return ctx.Error(err)
+			}
+		case payload.Empty:
+			if report.GetOrderedPendingCount() != 1 {
+				return ctx.Error(throw.New("Should be one ordered pending"))
+			}
+
+			objectPulse := report.GetAsOf()
+			currPulse := s.pulseSlot.PulseNumber()
+			orderedPendingPulse := report.GetOrderedPendingEarliestPulse()
+
+			if orderedPendingPulse < objectPulse || currPulse < orderedPendingPulse {
+				return ctx.Error(throw.New("Incorrect pending ordered pulse number"))
+			}
+		case payload.Inactive:
+			if err := validateZeroPending(report); err != nil {
+				return ctx.Error(err)
+			}
+			if err := validateEmptyState(report); err != nil {
+				return ctx.Error(err)
+			}
+		}
+	}
+
+	return ctx.Jump(s.stepProcess)
+}
+
+func validateZeroPending(report *payload.VStateReport) error {
+	if report.GetUnorderedPendingCount() != 0 ||
+		!report.GetUnorderedPendingEarliestPulse().IsUnknown() ||
+		report.GetOrderedPendingCount() != 0 ||
+		!report.GetOrderedPendingEarliestPulse().IsUnknown() {
+		return throw.New("")
+	}
+
+	return nil
+}
+
+func validateEmptyState(report *payload.VStateReport) error {
+	if report.ProvidedContent == nil {
+		return nil
+	}
+
+	switch content := report.ProvidedContent; {
+	case content.LatestValidatedState != nil:
```

Comment:
fallthrough? or you should use comma

Reply:
fallthrough, switch more simple that cond1 || cond2 || cond3 || .... expretion